src/tikv/Region.cc

```diff
@@ -0,0 +1,274 @@
+#include <common/Exception.h>
+#include <tikv/Region.h>
+
+namespace pingcap
+{
+namespace kv
+{
+
+RPCContextPtr RegionCache::getRPCContext(Backoffer & bo, const RegionVerID & id, bool is_learner)
+{
+    RegionPtr region = getCachedRegion(bo, id);
+    if (region == nullptr)
+    {
+        throw Exception("not found region, region id is: " + std::to_string(id.id));
+    }
+    const auto & meta = region->meta;
+    auto peer = region->peer;
+
+    if (is_learner)
+    {
+        peer = region->learner;
+        if (peer.store_id() == 0)
+        {
+            dropRegion(id);
+            throw Exception("no learner! the request region id is: " + std::to_string(id.id));
+        }
+    }
+
+    std::string addr = getStoreAddr(bo, peer.store_id());
+    if (addr == "")
+    {
+        dropRegion(id);
+        throw Exception("miss store, region id is: " + std::to_string(id.id) + " store id is: " + std::to_string(peer.store_id()));
+    }
+    return std::make_shared<RPCContext>(id, meta, peer, addr);
+}
+
+RegionPtr RegionCache::getCachedRegion(Backoffer & bo, const RegionVerID & id)
+{
+    std::lock_guard<std::mutex> lock(region_mutex);
```

Comment:
The lock could be hold for long time if `loadRegionByID` needs retry.

Reply:
Change it to RW lock